BParkingNano/plugins/ElectronMerger.cc

```diff
@@ -38,93 +50,134 @@ class ElectronMerger : public edm::global::EDProducer<> {
-
   edm::Handle<pat::ElectronCollection> lowpt;
   evt.getByToken(lowpt_src_, lowpt);
-
   edm::Handle<pat::ElectronCollection> pf;
   evt.getByToken(pf_src_, pf);
-  
-  std::unique_ptr<pat::ElectronCollection> out(new pat::ElectronCollection);
-
-  //just to avoid electrons double storing
-  //in case of ntrgMuon > 1 and same electron compatible with all 
-  std::vector<int> alreadySavedPF;
-  alreadySavedPF.resize(pf->size(), 0);
-
-  std::vector<int> alreadySavedLPT;
-  alreadySavedLPT.resize(lowpt->size(), 0);
-
-  for(auto muonTrg : *trgMuon) {
-
-    int icount = -1;
-    for(auto ele : *pf) {
-      ++icount;
-      if(alreadySavedPF[icount]) continue;
-
-      if((reco::deltaR(ele, muonTrg) < drTrg_cleaning_ && drTrg_cleaning_ != -1) ||
-	 (std::fabs(ele.vz() - muonTrg.vz()) > dzTrg_cleaning_ && dzTrg_cleaning_ != -1)) continue;
-
-      ele.addUserInt("isPF", 1);
-      ele.addUserInt("isLowPt", 0);
-      ele.addUserFloat("ptBiased", 20.);
-      ele.addUserFloat("unBiased", 20.);
-      ele.addUserFloat("mvaId", 20.);
-      ele.addUserFloat("chargeMode", ele.charge());
-      alreadySavedPF[icount] = 1;
-      out->push_back(ele);
-    }
+  edm::Handle<edm::ValueMap<float> > ptBiased;
+  evt.getByToken(ptBiased_src_, ptBiased);
+  edm::Handle<edm::ValueMap<float> > unBiased;
+  evt.getByToken(unBiased_src_, unBiased);
+  edm::Handle<edm::ValueMap<float> > mvaId;  
+  evt.getByToken(mvaId_src_, mvaId);
 
-    icount = -1;
-    for(auto ele : *lowpt) {
-      ++icount;
-      if(alreadySavedLPT[icount]) continue;
-
-      if((reco::deltaR(ele, muonTrg) < drTrg_cleaning_ && drTrg_cleaning_ != -1) ||
-	 (std::fabs(ele.vz() - muonTrg.vz()) > dzTrg_cleaning_ && dzTrg_cleaning_ != -1)) continue;
-
-      bool clean_out = false;
-      for(const auto& pfele : *pf) {
-
-	clean_out |= (
-		      fabs(pfele.vz() - ele.vz()) < dz_cleaning_ &&
-		      reco::deltaR(ele, pfele) < dr_cleaning_
-		      );
-      }
-      if(clean_out) continue;
-      ele.addUserInt("isPF", 0);
-      ele.addUserInt("isLowPt", 1);
-      ele.addUserFloat("chargeMode", ele.gsfTrack()->chargeMode());
-      if(use_gsf_mode_for_p4_) {
-	reco::Candidate::PolarLorentzVector p4(
-					     ele.gsfTrack()->ptMode(),
-					     ele.gsfTrack()->etaMode(),
-					     ele.gsfTrack()->phiMode(),
-					     ele.mass()
-					     );
-	ele.setP4(p4);
-      }
-      alreadySavedLPT[icount] = 1;
-      out->push_back(ele);
-    }
+  // output
+  std::unique_ptr<pat::ElectronCollection> ele_out(new pat::ElectronCollection);
 
-    std::sort(
-	      out->begin(), out->end(), 
-	      [] (pat::Electron e1, pat::Electron e2) -> bool {return e1.pt() > e2.pt();}
+  
+  // -> changing order of loops ert Arabella's fix this without need for more vectors  
+ for(auto ele : *pf) {
+   //cuts
+   if (ele.pt()<ptMin_) continue;
+   if (fabs(ele.eta())>etaMax_) continue;
+   // apply conversion veto unless we want conversions
+   if (!ele.passConversionVeto()) continue;
```

Comment:
I would suggest to avoid the proliferation of configurable cut values by putting a `StringCutObjectSelector` like [here](https://github.com/mverzett/BAnalysisCode/blob/nano_v2/ParkingNtuples/plugins/DiLeptonBuilder.h#L27)

Reply:
The problem with that approach is that it adds overhead time[*], and given that we want to run both on huge collections and to be fast at the same time applying everything in one loop is the most performing option. And since we are bound to have at least a loop for Dz cut and merging, is better to put everything there. Of course we can change it but that means that we are going to make more loops since Dz, which is necessary, cannot be apply in this Selector. Also cuts like min BDT in low pT e may become very tricky and cause one loop to evaluate, one to apply etc. Also it is a bit better in terms of bookkeeping since we now exactly which cuts will be used (they are configurable but IN our cfgs after all)


[*]What may cause a slight performance loss is the actual evaluation of expression. The parses, in fact, creates a composite structure in memory that invokes a virtual function call for every syntactical element in the expression.

Moreover, the passed object (track, in the example) has to be "converted" into a Reflex object in order to be evaluated.

Reply:
* The time overhead is really minimal, and no matter the size of the collection, the vertex creation will make it insignificant.
   *  No need for multiple loops, you can just 
      1. embed 
      2. evaluate cut parser
      3. check for DR and Dz

Reply:
Good morning Mauro,

Well in order to embed (1 loop) evaluate kinematic cuts (2 loops) finally apply DR Dz cuts (3 loops) is essentially 3 times the time needed right? Well it would be indeed insignificant if we were to create every possible candidate. But we apply cuts so a really small part of all possible triplets (10^-4 ) will be promoted, while the loops on the inputs will be there for every event constantly.

Reply:
@gkaratha the number of loops would be unaffected, `StringCutObjectSelector ` is nothing but a function that evaluates to a bool that you can use in an if statement, you just replace the condition.

I forgot your comment regarding the reflex dictionaries. We (CMS) build reflex dictionaries for every object we store in the event (ROOT needs them) so they are already there, and anyhow those are done at compilation time, not at runtime: that's the reason why compiling the modules in DataFormats takes so long.

Reply:
@mverzett , There is also the time reading and writing the new products which at least embeder and Dz cut will have (btw what is DR cut?). So even if selector makes no loops still the embeder and Dz will have. So everything will be 2 times slower, no matter how small that is. Now, no matter how small the total overhead per evt is, it is something that we will "pay" in every evt **for 10^10 evts** while fitter after cuts will be invoked for a small number of events. **So the end effect is not clear to me**. Also, at least in my point of view, is better to have a single producer per object, concentrating everything about the object. An additional point is that when we introduce muon+track and keep the same logic(embeder+selector+Dz) it will be more overhead in a code that is not very fast to begin with.  In any case i did not removed any of the files, just comment them out so they can be commented in :) 

Bottom line: The logic behind this change is to have the fastest code possible, and from the conversation I dont see any objection on that. Now if we want to have this or this+Selector or the embeder+selector+this is something we can discuss ;)